### PR TITLE
Flatten schema pkg

### LINKED ISSUE #4534: schema: redesign v* package layout
Currently the `schema` module has the following package layout

```text
schema
├── internal
├── v1.0
│   ├── ast
│   ├── testdata
│   └── types
└── v1.1
    ├── ast
    ├── testdata
    └── types
```

Using packages with `v*` [has semantic meaning in Go](https://go.dev/blog/v2-go-modules).

Instead of using the `v1.0` and `v1.1` directory structure to partition the supported schema file versions, look into alternatives so if in the future there needs to be a `v2` version of the `schema package we can accommodate this.

### PR DESCRIPTION
**Pull request by MrAlias**
Fixes #4534

- Flattens the `schema/v1.0` and `schema/v1.1` packages into `schema`.
- A new  approach is taken to file format compatibility:
  - Have the `schema` package parse the latest file format and all versions less than that into the same struct
  - When the file format introduces incompatible changes, introduce a `v2` of `schema`
- Don't use a type declaration when an equivalent built-in type will do
- Add the `All` and `Resource` types to replace the generic `Attributes` type that was used by both
  - This ensures that if one of these fields changes in the future the change can be independently made
- Rename `VersionDef` to `Changeset` to better match its purpose as the differences between two successive versions

### CODE REVIEW

File: schema/parser.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+// Parse parses a Schema read from r.
+//
+// If r contains a Schema with a file format version higher than FileFormat, an
+// error will be returned.
+//
+// If r contains an invalid schema URL an error will be returned.
+func Parse(r io.Reader) (*Schema, error) {
+	d := yaml.NewDecoder(r)
+	d.KnownFields(true)
```
[the remark this answers is not in the document]
Comment (MrAlias) [reply schema/parser.go:48]: I had considered restricting parsing to only allow parsing of a specific version of the file format, but I don't think that's the user experience we want to provide. If a user has a schema that incorrectly states it is 1.0.0 and uses 1.1.0 fields, I would rather produce a valid schema for the user. I doubt they care that the schema was incorrect, they just want the data.

[the remark this answers is not in the document]
Comment (pellared) [reply schema/parser.go:48]: I agree. Should we add tests for this behavior and should it be documented?

[the remark this answers is not in the document]
Comment (MrAlias) [reply schema/parser.go:48]: > I agree. Should we add tests for this behavior and should it be documented?

Yeah, that sounds good to me. I'll update in a bit.

[the remark this answers is not in the document]
Comment (MadVikingGod) [reply schema/parser.go:48]: There is another section in the file format that specifies that this is the intended behavior:

> Correspondingly:
> 
>     Consumers of the schema file SHOULD NOT attempt to interpret the schema file if the MAJOR version number is different (higher or lower) than what the consumer supports.
> 
>     Consumers of the schema file SHOULD NOT attempt to interpret the schema file if the MINOR version number is higher than what the consumer supports.
> 
>     Consumers MAY ignore the PATCH number.

File: schema/parser.go [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+	var s Schema
+	if err := d.Decode(&s); err != nil {
```
Comment (MadVikingGod) [inline schema/parser.go:51]: Current behavior is that all version fields (`1.1.0` etc) will be returned in the order they are in the file.  

The format implies that they should be in the order of highest semver to lowest.  Do we want to take any actions to put them in that order?

**Review by MadVikingGod — APPROVED** [review]
A README.md would be useful.

**Comment by pellared** [comment]
> A README.md would be useful.

`schema/parser.go` contains package documentation. We could move it to `schema/doc.go`.

**Comment by MrAlias** [comment]
Closing. Not needed: https://github.com/open-telemetry/opentelemetry-go/pull/4503#issuecomment-1918175963

**Pull request closed without merging**